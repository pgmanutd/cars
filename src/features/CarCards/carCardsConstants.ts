export const CAR_CARDS_FOR_LOADER_COUNT = 5;
