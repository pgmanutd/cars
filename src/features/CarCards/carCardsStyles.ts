import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    carImageContainer: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    carImage: {
      height: 80,
      width: 80,
    },
    carContainer: {
      border: `1px solid ${theme.palette.divider}`,
      padding: theme.spacing(2),
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(2),
    },
  }),
);
