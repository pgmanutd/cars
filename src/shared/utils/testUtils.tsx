import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {},
) => ({
  ...render(<Router history={history}>{ui}</Router>),
  history,
});
