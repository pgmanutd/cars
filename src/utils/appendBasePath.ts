const appendBasePath = (basePath: string, path: string) =>
  `${basePath}?${path}`;

export default appendBasePath;
