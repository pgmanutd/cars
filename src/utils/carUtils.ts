import { Car } from 'features/CarInfo/types';

export const getCarName = ({ manufacturerName, modelName }: Car) =>
  `${manufacturerName} ${modelName}`;

export const getCarFeatures = (
  { stockNumber, mileage, fuelType, color }: Car,
  { stockNumberPrefix = '' }: { stockNumberPrefix?: string } = {
    stockNumberPrefix: '',
  },
) =>
  `${stockNumberPrefix}${stockNumber} - ${mileage?.number} ${mileage?.unit} - ${fuelType} - ${color}`;
