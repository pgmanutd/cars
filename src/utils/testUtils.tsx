import React from 'react';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    routerConfig: {
      path = '/',
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    },
  }: {
    routerConfig: { path?: string; route?: string; history?: MemoryHistory };
  } = { routerConfig: {} },
) => ({
  ...render(
    <HelmetProvider>
      <Router history={history}>
        <Route
          path={path}
          render={routeProps => React.cloneElement(ui, { ...routeProps })}
        />
      </Router>
    </HelmetProvider>,
  ),
  history,
});
