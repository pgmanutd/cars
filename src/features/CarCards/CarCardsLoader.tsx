import React from 'react';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import Skeleton from '@material-ui/lab/Skeleton';

import { CAR_CARDS_FOR_LOADER_COUNT } from './carCardsConstants';
import { useStyles } from './carCardsStyles';

const CarCardsLoader: React.FC = () => {
  const classes = useStyles();

  return (
    <Box data-testid="CarCardsLoader" component="section">
      {Array(CAR_CARDS_FOR_LOADER_COUNT)
        .fill({})
        .map((_, index) => (
          <Grid
            data-testid="CarCardLoader"
            key={index}
            component="section"
            container
            spacing={2}
            className={classes.carContainer}
          >
            <Grid item xs={2}>
              <Skeleton height={85} width={100} variant="rect" />
            </Grid>
            <Grid item xs={5}>
              <Skeleton height={32} width="80%" />
              <Skeleton height={20} width="80%" />
              <Skeleton height={16} width="30%" />
            </Grid>
            <Grid item xs={5}>
              <Skeleton height={150} variant="rect" />
            </Grid>
          </Grid>
        ))}
    </Box>
  );
};

export default CarCardsLoader;
