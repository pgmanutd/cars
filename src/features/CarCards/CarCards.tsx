import React, { memo } from 'react';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import CardMedia from '@material-ui/core/CardMedia';
import Link from '@material-ui/core/Link';
import Typography from '@material-ui/core/Typography';
import { Link as RouterLink } from 'react-router-dom';

import routePaths from 'constants/routePaths';

import { getCarName, getCarFeatures } from 'utils/carUtils';

import { useTranslate } from 'features/Translate';
import ToggleFavoriteCar from 'features/ToggleFavoriteCar';

import { Car } from 'features/CarInfo/types';

import CarCardsLoader from './CarCardsLoader';
import { useStyles } from './carCardsStyles';

export interface CarCardsProps {
  cars: Car[];
  isLoading?: boolean;
}

const CarCards: React.FC<CarCardsProps> = ({
  cars,
  isLoading = false,
  ...restProps
}) => {
  const classes = useStyles();
  const { translate } = useTranslate();

  if (isLoading) {
    return <CarCardsLoader />;
  }

  return (
    <Box data-testid="CarCards" {...restProps} component="section">
      {cars.map(car => (
        <Grid
          data-testid="CarCard"
          {...restProps}
          key={car.stockNumber}
          component="section"
          container
          spacing={2}
          className={classes.carContainer}
        >
          <Grid item md={2} xs={12}>
            <Box className={classes.carImageContainer}>
              {car.pictureUrl ? (
                <CardMedia
                  image={car.pictureUrl}
                  title={car.modelName}
                  className={classes.carImage}
                />
              ) : (
                car.modelName
              )}
            </Box>
          </Grid>
          <Grid item md={5} xs={12}>
            <Typography variant="h6" gutterBottom>
              {getCarName(car)}
            </Typography>
            <Typography variant="body2" gutterBottom>
              {getCarFeatures(car, {
                stockNumberPrefix: translate(
                  'features.CarCards.stockNumberPrefix',
                ),
              })}
            </Typography>
            <Link
              component={RouterLink}
              to={routePaths.carDetails.pathWithParams({
                stockNumber: car.stockNumber?.toString(),
              })}
              color="textSecondary"
              variant="body2"
            >
              {translate('features.CarCards.viewDetailsLinkText')}
            </Link>
          </Grid>
          <Grid item md={5} xs={12}>
            <ToggleFavoriteCar car={car} />
          </Grid>
        </Grid>
      ))}
    </Box>
  );
};

// NOTE: Using memo to avoid unnecessary rerenders of this component.
// Only used "memo" for components which actually accepts any props and used
// it for rendering
// https://kentcdodds.com/blog/usememo-and-usecallback
export default memo(CarCards);
