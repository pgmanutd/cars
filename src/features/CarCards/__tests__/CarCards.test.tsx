import React from 'react';

import { renderWithProviders } from 'utils/testUtils';

import CarCards from '../CarCards';

describe('<CarCards />', () => {
  const setup = ({ isLoading = false } = {}) => {
    const cars = [
      {
        stockNumber: 1234,
        manufacturerName: 'Audi',
        modelName: 'Abcd',
        mileage: {
          number: 1234,
          unit: 'km',
        },
        fuelType: 'petrol',
        color: 'white',
        pictureUrl: 'image.svg',
      },
    ];
    const { renderResult } = renderWithProviders(
      <CarCards cars={cars} isLoading={isLoading} />,
    );

    return {
      renderResult,
    };
  };

  it('should render the component and matches it against stored snapshot', () => {
    const { renderResult } = setup();

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should render <CarCardsLoader /> when #isLoading prop is true', () => {
    const { renderResult } = setup({ isLoading: true });

    expect(renderResult.getByTestId('CarCardsLoader')).toBeInTheDocument();
  });
});
