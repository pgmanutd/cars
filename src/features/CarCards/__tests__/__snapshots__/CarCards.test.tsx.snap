// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarCards /> should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-5"
    data-testid="CarCards"
  >
    <section
      class="MuiGrid-root makeStyles-carContainer-3 MuiGrid-container MuiGrid-spacing-xs-2"
      data-testid="CarCard"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-2"
      >
        <div
          class="MuiBox-root MuiBox-root-109 makeStyles-carImageContainer-1"
        >
          <div
            class="MuiCardMedia-root makeStyles-carImage-2"
            style="background-image: url(image.svg);"
            title="Abcd"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-5"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
        >
          Audi Abcd
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
        >
          Stock #1234 - 1234 KM - Petrol - White
        </p>
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorTextSecondary"
          href="/details/1234"
        >
          View details
        </a>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-5"
      >
        <section
          class="MuiBox-root MuiBox-root-151 makeStyles-root-149"
          data-testid="ToggleFavoriteCar"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            If you like this car, click the button and save it in your collection of favorite items.
          </p>
          <div
            class="MuiBox-root MuiBox-root-152 makeStyles-buttonContainer-150"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Save
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </section>
      </div>
    </section>
  </section>
</DocumentFragment>
`;
